sound/soc/sof/ipc4-topology.c

```diff
@@ -2015,6 +2056,10 @@ static int sof_ipc4_route_setup(struct snd_sof_dev *sdev, struct snd_sof_route *
 				      SOF_PIN_TYPE_SINK);
 	}
 
+	if (sroute->src_queue_id > 0 && WIDGET_IS_COPIER(src_widget->id))
+		sof_ipc4_set_copier_sink_format(sdev, src_widget, sink_widget,
+						sroute->src_queue_id);
```

Comment:
@bardliao can you elaborate on why we don't need to setup the format for the src_queue_id 0? where is this done? Why can't we have a single place where all formats are setup for all source queues?

Reply:
> @bardliao can you elaborate on why we don't need to setup the format for the src_queue_id 0? where is this done? Why can't we have a single place where all formats are setup for all source queues?

@plbossart Pin 0 format is included in `struct sof_ipc4_copier_data {}`  with MOD_INIT_INSTANCE IPC.
```
struct sof_ipc4_copier_data {
        struct sof_ipc4_base_module_cfg base_config;
        struct sof_ipc4_audio_format out_format;
        uint32_t copier_feature_mask;
        struct sof_copier_gateway_cfg gtw_cfg;
};
```
In other words, pin 0 format is set when the copier widget init. But any additional pin format should be set by LARGE_CONFIG_SET IPC message. We can't set the non-zero pin formats during setup stage since the format need to be set after binding. See @RanderWang's comment https://github.com/thesofproject/linux/pull/4006#discussion_r1019841908.
Also, we won't know what widget is connected to the output pin and what format it needs before the route is setup.

Reply:
Copier can be added sink module dynamically so it is fit for doing it at binding time